pkg/migration-tool/writer/shards.go

```diff
@@ -0,0 +1,181 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+// shardsSet represents a set of shards. It consists of configuration related
+// to handling and management of shards.
+type shardsSet struct {
+	num         int
+	set         []*shard
+	cancelFuncs []context.CancelFunc
+
+	done    chan struct{}
```

Comment:
optional: a more idiomatic way to do this is to use a sync.WaitGroup, I'll leave it up to you if you want to change it.

Reply:
The reason why wait group is not used is that using wait-group will not give the signal on the writer goroutine when a shard is complete in order to proceed the progress-bar per shard level, rather just give a whole done, making the progress-bar fewer steps (hence less interactive).